File: Languages/python/algorithms-dsa/leetcode/greedy/partition_labels.py
"""
LeetCode #763 - Partition Labels
Topic: Greedy
Difficulty: Medium

Partition string into maximum parts so each letter appears in at most one part.
Track last occurrence of each character.

Time Complexity: O(n)
Space Complexity: O(1) - at most 26 keys
"""


def partition_labels(s: str) -> list[int]:
    last: dict[str, int] = {ch: i for i, ch in enumerate(s)}
    partitions: list[int] = []
    start = end = 0
    for i, ch in enumerate(s):
        end = max(end, last[ch])
        if i == end:
            partitions.append(end - start + 1)
            start = i + 1
    return partitions


if __name__ == "__main__":
    assert partition_labels("ababcbacadefegdehijhklij") == [9, 7, 8]
    assert partition_labels("eccbbbbdec") == [10]
    assert partition_labels("abc") == [1, 1, 1]
    print("All tests passed!")

File: Languages/cpp/algorithms-dsa/leetcode/greedy/partition_labels.cpp
/**
 * LeetCode 763: Partition Labels
 * Topic: Greedy
 * Difficulty: Medium
 *
 * Partition a string so each letter appears in at most one part.
 * Track last occurrence of each character, extend partition boundary.
 * Time: O(n), Space: O(1)
 */
#include <iostream>
#include <vector>
#include <string>
#include <cassert>
using namespace std;

vector<int> partitionLabels(string s) {
    vector<int> last(26, 0);
    for (int i = 0; i < (int)s.size(); i++)
        last[s[i] - 'a'] = i;

    vector<int> res;
    int start = 0, end = 0;
    for (int i = 0; i < (int)s.size(); i++) {
        end = max(end, last[s[i] - 'a']);
        if (i == end) {
            res.push_back(end - start + 1);
            start = i + 1;
        }
    }
    return res;
}

int main() {
    auto r1 = partitionLabels("ababcbacadefegdehijhklij");
    assert(r1 == vector<int>({9, 7, 8}));

    auto r2 = partitionLabels("eccbbbbdec");
    assert(r2 == vector<int>({10}));

    auto r3 = partitionLabels("abc");
    assert(r3 == vector<int>({1, 1, 1}));

    cout << "All tests passed!" << endl;
    return 0;
}
